### ISSUE TITLE
ZSON: accept omitted nulls

### ISSUE THREAD

**Issue by henridf** [brimsec/zq#2006]
Currently, ZSON requires null fields to be explicit, including when the record type is known.  For example, given 
```
{
    a: 0,
    b: 0,
} (=1)
```
then `{a: 0} (1)` (currently invalid due to missing field `b`) would be identical to `{ a: 0, b: null }`.

This issue consists of changing the spec and implementation to handle implicit nulls in this way. It is a backwards-compatible change.

To be determined if implicit nulls should be zq's default behavior when emitting ZSON, or enabled by a flag, etc.

This was pointed out by @aousterh.

**Comment by mccanne** [comment]
ZSON is just a format and I'm not sure it support casting like this.  You can do it in the Zed language...
```
cat in.zson
{a:0,b:0}(=x)
{a:0}
% zq -z "yield cast(this,<{a:int64,b:int64}>)" in.zson
{a:0,b:0}
{a:0}
% zq -z "yield shape(this,<{a:int64,b:int64}>)" in.zson
{a:0,b:0}
{a:0,b:null(int64)}
```
(I was surprised cast didn't work here but shape does... is this documented?)

Moving this to backlog for flag week to fix or close any of these issues here...

**Comment by philrz** [comment]
I happened to revisit this one during a shaping issue scrub. Here's a summary of how things look at current Zed commit 8bf604c.

Combining @henridf's original observation with @mccanne's variations in his comment, here's how the error looks when triggered:

```
$ cat in-err.zson 
{a:0,b:0}(=x)
{a:0}(x)

$ zq -version
Version: v1.8.1-53-g8bf604cf

$ zq -z in-err.zson 
in-err.zson: record decorator fields (2) mismatched with value fields (1)
```

@mccanne's findings about how it can be accomplished in the language still hold.

```
$ cat in.zson 
{a:0,b:0}(=x)
{a:0}

$ zq -z "yield cast(this,<{a:int64,b:int64}>)" in.zson
{a:0,b:0}
{a:0}

$ zq -z "yield shape(this,<{a:int64,b:int64}>)" in.zson
{a:0,b:0}
{a:0,b:null(int64)}

$ zq -z "yield shape(this,<x>)" in.zson
{a:0,b:0}(=x)
{a:0,b:null(int64)}(=x)
```

As for @mccanne's question about if `cast()`'s behavior here is documented, indeed it is. For instance, [here](https://zed.brimdata.io/docs/next/language/shaping#cast) it's mentioned in brief how "each leaf value that matches the field path in the specified type" (i.e., implying that non-matching field paths are not touched), and in more detail in the full [`cast` function docs](https://zed.brimdata.io/docs/next/language/functions/cast) where it's explained that "absent fields are ignored and omitted from the result". So, I think we're covered there.

Looking at the overall subject with fresh eyes, I'm inclined to agree with @mccanne's assessment that cast-like behavior within the ZSON stream is probably not necessary. It doesn't seem like the Zed tooling/libraries themselves would ever need to create ZSON that would rely on this functionality were it to exist. In general it seems like we should expect that ZSON is being constructed by the Zeek tooling/libraries, i.e., users crafting ZSON by hand may certainly happen, but it's questionable if there should be extra code that accommodates lax shorthand such as what's implied here.

In conclusion, I suspect this issue could be closed, but it's probably not my place to make that call on my own. Since the original observation was attributed to @aousterh and I know she continues to keep an eye on the projects, perhaps she'd be willing to chime in with a current take.

**Comment by aousterh** [comment]
Thanks for looping me in @philrz.

I started thinking about this aspect of ZSON because I was looking at sparse data, where an individual record might have many `null` fields in it. For that kind of data, I thought it would be useful to have a way to omit those `null`s. As with JSON, this could make it easier for users to read sparse ZSON, and easier to craft ZSON by hand (though I agree that this latter case may not be very common).

If the decorator `(x)` is valid syntax then I would expect it to have a different meaning from `(=x)` and to mean that the record has type `(x)`. In that case, it seems unnecessary to specify all of the `null` fields of the record, since the type is already explicitly specified. I like @henridf's suggestion to have a flag that causes `null` fields to be omitted when ZSON is output, using the `(x)` syntax.
